### ISSUE TITLE
Refactor PASSWORD_DIALOG

### ISSUE THREAD

**Issue by larshp** [larshp/abapGit#783]
related to #773 

Requirements:
* No use of selection screen or custom dynpros, it should work with global classes
* Decouple GUI and logic, `lcl_password_dialog=>popup` is called deep inside the http logic

Plan:
A1: add optional username+password to the http class
A2: if username+password is not supplied raise resumeable(?) exception

B1: add new `page_password` for entering username+password
B2: some way for calling `page_password` and returning username+password

**Comment by sbcgua** [comment]
it can be commit or private repo. I guess strategy A would be better. Better UX + native SAP memory for field history (if several users). But completely without dynpros ?? could be difficult. Some way dynamic dynpro creation ? or dynamically create and save dynpro to ZABAPGIT ? Or find some existing screen for the purpose (but I failed when I was up to this feature - there are some standard screens for password request but they are really ugly)

**Comment by larshp** [comment]
Want to do both A and B, it is just 2 disconnected tasks.

Dynpros: want to create a lcl_gui_page_password html page, it can look nice and have all the needed information

**Comment by marcellourbani** [comment]
Looking into this right now due to a demo project I'm working on.
I think calling a form is really bad, and passing this stuff down is also messy.

For my own stuff I usually use a rudimentary container/generic factory which works more or less like this:
```
data: identityservice type ref to zif_identityservice.

identityservice ?= container->get( 'zif_identityservice' ).

userdata = identityservice->getuserdata( ).
```
The container has a default implementer (say zcl_whatever for zif_whatever) which can be overridden by providing either an object or a class name. 

In here I didn't see any such mechanism. Even a simpler one , like replacing 

`zcl_abapgit_password_dialog=>popup( ).`

with 

`zcl_abapgit_password_dialog=>get_instance( )->popup( ).`

which would allow a more ad-hoc implementation like 

`zcl_abapgit_password_dialog=>set_instance( my_override )`

in the main caller
Also I don't like this bit in the popup class:

```
    IF lv_gui_is_available = abap_true.
      PERFORM password_popup
        IN PROGRAM (sy-cprog)
        USING iv_repo_url
        CHANGING cv_user cv_pass.
    ELSE.
...
          CALL METHOD ('ZCL_ABAPGIT_DEFAULT_AUTH_INFO')=>('GET_USER')
            RECEIVING
              rv_user = cv_user.
...
    ENDIF.
```
which looks like a patch over the inability to provide an user service dynamically

Any thoughts?

**Comment by larshp** [comment]
depends on the scenario, will setting username and password via `zcl_abapgit_login_manager` work?

**Comment by sbcgua** [comment]
Guess the identity-instance factory would fit somewhere in `zcl_abapgit_environment`, it also knows about GUI presence so would not need `set` (for now :).
`zcl_abapgit_login_manager` should be quite independent - it is called prior to the dialog (http->create_by_url).
NB: password_dialog also used in `ZCL_ABAPGIT_PROXY_AUTH`

**Comment by marcellourbani** [comment]
login manager works fine. Thank you guys

Relatively easy to spot too with hindsight :man_facepalming: 
Would probably like better a design where the popup is an alternative implementation of the login manager but not an issue once the contract is clear.

It's more an aesthetic thing: don't like to put any UI/popups in services unless explicitly injected

**Comment by marcellourbani** [comment]
for the record, we also have an injector, I used it to get rid of another popup:

```
create object dummy.
 zcl_abapgit_ui_injector=>set_gui_functions( dummy ).```

I still think  we should refactor to put all popups in a service, so creating a non interactive one would be more consistent

**Comment by larshp** [comment]
yes, backend should be decoupled from the user interface, one step at a time

**Comment by larshp** [comment]
also see comments in https://github.com/larshp/abapGit/issues/1331

**Comment by mbtools** [comment]
How about this?

![image](https://user-images.githubusercontent.com/59966492/96312976-b7733500-0fda-11eb-8510-89a657bd5479.png)

Best thing about it: it's not a fake but working!

Worst thing about it: to be a clean solution, it uses a new `zcx_abapgit_login_error` exception that would have to be added to many classes (the raise is in `zcl_abapgit_http->create_by_url` and the catch is in `zcl_abapgit_gui->handle_action`).

**Comment by flaiker** [comment]
> How about this?

How can a password manager autofill these fields? They rely on the window title which is unique in the popup for each service and also on the input field focus.

**Comment by sbcgua** [comment]
Looks nice. A couple of thoughts:
1) probably would be nice to indicate the repo in the title and the operation performed - which is the context for giving the credentials. **If** this is separate page ...
2) Speaking of which ... would be nice to have the dialog on the same page where the operation was requested. If possible ... this is not easy to do it without 2-sided communication unfortunately.

**Comment by mbtools** [comment]
- It now sets the GUI title similar to the old popup to "<Sysid> - Login: <host>":

![image](https://user-images.githubusercontent.com/59966492/96650227-72177600-1300-11eb-89ca-feb7ef10390d.png)

- Gets/sets default user 
- Input focus of in user fields (no default user) or password (with default user)
- Shows repo name in page title (but not in GUI title)
- Initial message "You have to login to access this repository" (like warning)
- Subsequent messages "Sorry, username or password are not correct. Try again!" (like error)

![image](https://user-images.githubusercontent.com/59966492/96650322-ab4fe600-1300-11eb-8a7e-14c61300d500.png)

- Up to 3 login attempts. Afterwards system goes to target page and displays error (as before)

Open issues:
1. Should GUI title contain system id (as in screenshot)?
2. Although form focus is set to correct field, SAP GUI does not initially focus on the HTML control. Any idea what needs to change?
3. Login using "Digest" has to be added again (wip). 
4. Dialog on same page: That will be tricky. We can try after this one works.

**Comment by flaiker** [comment]
> Should GUI title contain system id as a prefix ("M0P" in screenshot)?

No, this is already covered by the SAP GUI setting "Show system name in taskbar button".

**Comment by larshp** [comment]
from https://github.com/abapGit/abapGit/pull/4052,

> Suggest the following 3 PRs:
> 
> refactoring logic in zcl_abapgit_login_manager, digest and acquiring login details
> refactoring the login handling with the exception can be a separate PR, while still keeping the old popup
> refactoring the old popup to new HTML
